### ISSUE TITLE
Stack Level Too Deep in unpack_compressed

### ISSUE THREAD

**Issue by masukomi** [mojombo/grit#67]
grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:314: stack level too deep (SystemStackError)

the referenced line is 
            zstr = Zlib::Inflate.new

I can provide more details if you tell me what you need.
Judging by the comments in issue #41 this is not a duplicate.

**Comment by bobbywilson0** [comment]
It would be helpful to describe what you were doing to create this issue, also what your are trying to get done. Sometimes there is another way to do it. Include relevant code.

**Comment by masukomi** [comment]
sorry about the delay. 
the offending line of code:

```
branch_commits = repo.commits_since(branch.name, @options[:since] ) 
```

@options[:since] is "Thu Jun 16 00:00:00 -0400 2011"
branch is one of the items from repo.heads() like "master"

the stack trace:

```
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:200:in `find_object_in_index'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:233:in `find_object'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:70:in `with_idx'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:232:in `find_object'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/internal/pack.rb:115:in `[]'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:78:in `get_raw_object_by_sha1'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:77:in `each'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:77:in `get_raw_object_by_sha1'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:105:in `get_object_by_sha1'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:372:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `each'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:376:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `each'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:376:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `each'
    --- the above 3 lines repeated MANY times ----
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:371:in `walk_log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:287:in `log'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby/repository.rb:306:in `rev_list'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby.rb:68:in `rev_list'
/usr/local/rvm/gems/ruby-1.8.7-p334/gems/grit-2.4.1/lib/grit/git-ruby.rb:224:in `try_run'
/usr/local/rvm/rubies/ruby-1.8.7-p334/lib/ruby/1.8/timeout.rb:67:in `timeout'
```

**Comment by masukomi** [comment]
i should note that it doesn't always happen. every now and then it just hits an offensive branch in a particular repo... sadly it's "master" and not a public repo that you could test on. if i run something roughly equivalent on the command line

```
git log --since="Thu Jun 16 00:00:00 -0400 2011" --format=oneline
```

I only get 9 commits. Just checking on the off chance it was somehow related to a large # of commits (doubtful since it's a stack depth issue). 

I'm happy to run anything that would help diagnose this.

**Comment by bobbywilson0** [comment]
I actually noticed a few issues with the same `commits_since` method. I submitted a patch and it is still in the queue. I don't know what will happen with it, but I would appreciate it if you wanted to test my patch (https://bobbyw@github.com/bobbyw/grit.git) against your code to see if it fixes your situation. I worked specifically on the `commits_since` method.

**Comment by masukomi** [comment]
Commit 
    cdbb0f534ae615378bcfbfeb07cc2aef7e258b76 (in your branch... ignore the link)
    fixed commits_since on repo
 does not fix _this_ problem. I also tested your branch with all of your commits and the bug was still there.

**Comment by bobbywilson0** [comment]
Ah, I should have noticed this earlier, is @options[:since] a String or a Time object? Part of what I was working on was that the documentation and the default parameter are a String, but later in `commits_since` logic it checks to make sure that what was passed in is a Time object.

**Comment by masukomi** [comment]
it's a time object 
    options[:since] = Time.parse((Date.today() - options[:days]).to_s)

These errors are being generated via the git-status-report tool in https://github.com/masukomi/git_accessories if you want to attempt to replicate. Simplest way would be to cd into a git repo and run it with -a option. Won't require configuring anything. Full usage instructions are in the readme. 

75: options[:since] = Time.parse((Date.today() - options[:days]).to_s)
257: branch_commits = repo.commits_since(branch.name, @options[:since] )

sadly it takes a very particular repo / branch to actually make the error show up. Used it for over a year without problem. had to move to a new company to find a repo that could make this happen.

**Comment by bobbywilson0** [comment]
Do you happen to know of an open-source repo that will raise the error? I am interested in troubleshooting this with you but I feel like it will take me a while to even find a repo that raises the error you are getting.

**Comment by masukomi** [comment]
Alas, I do not. I'd been using that tool for quite some time without hitting that problem. I'll give it a whack on the linux kernel. if there's anything big and complex enough to trigger it I bet it's that.

**Comment by bobbywilson0** [comment]
I was trying to help debug #74 and he was using this repo: git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-2.6.38.y.git

I just out of curiosity I tried your git-status-report tool using the linux kernel and was able to replicate your issue! I managed to get a bit better output with rubinius https://gist.github.com/1081288. I had to truncate some of the output as the original paste was too large, but I don't think it is missing anything crucial. I actually think that this is related to issue #2, which is basically grit swallowing a non-zero exit code but not reporting it. It then just times out because it can't get passed that error.

CC @rhaamo
